presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -88,401 +110,746 @@
-                (Expression) process(node.getRight(), context));
+        Query oldQuery = node.getQuery();
+        Query newQuery = processSameType(oldQuery);
+        if (oldQuery == newQuery) {
+            return node;
+        }
+        if (node.getLocation().isPresent()) {
+            return new TableSubquery(node.getLocation().get(), processSameType(newQuery));
+        }
+        return new TableSubquery(processSameType(newQuery));
     }
 
     @Override
-    protected Node visitIdentifier(Identifier node, Void context)
+    protected Node visitAliasedRelation(AliasedRelation node, Void context)
     {
-        if (!materializedViewInfo.getBaseToViewColumnMap().containsKey(node)) {
-            throw new IllegalStateException("Materialized view definition does not contain mapping for the column: " + node.getValue());
+        Relation oldRelation = node.getRelation();
+        Relation newRelation = processSameType(oldRelation);
+        if (oldRelation == newRelation) {
+            return node;
         }
-        return new Identifier((materializedViewInfo.getBaseToViewColumnMap().get(node)).getValue(), node.isDelimited());
+        if (node.getLocation().isPresent()) {
+            return new AliasedRelation(
+                    node.getLocation().get(),
+                    processSameType(newRelation),
+                    node.getAlias(),
+                    node.getColumnNames());
+        }
+        return new AliasedRelation(
+                processSameType(newRelation),
+                node.getAlias(),
+                node.getColumnNames());
     }
 
     @Override
-    protected Node visitExpression(Expression node, Void context)
+    protected Node visitSampledRelation(SampledRelation node, Void context)
     {
-        return super.visitExpression(removeExpressionPrefix(node, removablePrefix), context);
+        Relation oldRelation = node.getRelation();
+        Relation newRelation = processSameType(oldRelation);
+        if (oldRelation == newRelation) {
+            return node;
+        }
+        if (node.getLocation().isPresent()) {
+            return new SampledRelation(
+                    node.getLocation().get(),
+                    newRelation,
+                    node.getType(),
+                    node.getSamplePercentage());
+        }
+        return new SampledRelation(
+                newRelation,
+                node.getType(),
+                node.getSamplePercentage());
     }
 
     @Override
-    protected Node visitFunctionCall(FunctionCall node, Void context)
+    protected Node visitJoin(Join node, Void context)
     {
-        if (!SUPPORTED_FUNCTION_CALLS.contains(node.getName())) {
-            throw new SemanticException(NOT_SUPPORTED, node, node.getName() + " function is not supported in query optimizer");
+        Relation oldLeft = node.getLeft();
+        Relation oldRight = node.getRight();
+        Relation newLeft = processSameType(oldLeft);
```

Comment:
I don't follow, the only way I'm interpreting this is `process(oldLeft)` instead of `processSameType(oldLeft)`? If so, what is unique about this method vs all the others where I'm calling `processSameType`?